File: bench/cpp/ops/BenchCenterCrop.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchCenterCropConfig.hpp"

#include <cvcuda/OpCenterCrop.hpp>
#include <cvcuda/OpReformat.hpp>

#include <nvbench/nvbench.cuh>

template<typename T>
inline void centercrop(nvbench::state &state, nvbench::type_list<T>)
try
{
    long3                       srcShape  = benchutils::GetShape<3>(state.get_string("shape"));
    auto                        layout    = benchutils::GetStringParam(state, "layout", "NHWC");
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));

    using BT = typename nvcv::cuda::BaseType<T>;

    int ch = nvcv::cuda::NumElements<T>;

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("CenterCrop benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    // CenterCrop is tensor-only; the native planar (NCHW) and fake-planar (NCHW_FAKE) paths are too.
    if ((isPlanar || isFakePlanar) && inputKind != benchutils::InputKind::Tensor)
    {
        state.skip("Planar CenterCrop benchmark is tensor-only");
        return;
    }

    nvcv::Size2D cropSize;

    if (state.get_string("cropType") == "SAME")
    {
        cropSize = nvcv::Size2D{(int)srcShape.z, (int)srcShape.y};
    }
    else if (state.get_string("cropType") == "QUARTER")
    {
        cropSize = nvcv::Size2D{(int)srcShape.z / 2, (int)srcShape.y / 2};
    }
    else
    {
        throw std::invalid_argument("Invalid cropType = " + state.get_string("cropType"));
    }

    long3 dstShape{srcShape.x, cropSize.h, cropSize.w};

    const long fullBytes = srcShape.x * srcShape.y * srcShape.z * sizeof(T);
    const long cropBytes = dstShape.x * dstShape.y * dstShape.z * sizeof(T);
    if (isFakePlanar)
    {
        // reformat(NCHW→NHWC) + crop + reformat(NHWC→NCHW).
        state.add_global_memory_reads(fullBytes + 2 * cropBytes);
        state.add_global_memory_writes(fullBytes + 2 * cropBytes);
    }
    else
    {
        state.add_global_memory_reads(cropBytes);
        state.add_global_memory_writes(cropBytes);
    }

    cvcuda::CenterCrop op;

    // clang-format off

    if (isFakePlanar) // tensor-only: planar→interleaved→crop→interleaved→planar
    {
        nvcv::Tensor src     ({{srcShape.x, ch, srcShape.y, srcShape.z}, "NCHW"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interSrc({{srcShape.x, srcShape.y, srcShape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interDst({{dstShape.x, dstShape.y, dstShape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst     ({{dstShape.x, ch, dstShape.y, dstShape.z}, "NCHW"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        cvcuda::Reformat reformatOp;

        benchutils::warmup_and_exec(state, BENCH_CENTERCROP_WARMUP_ITERATIONS,
            [&op, &reformatOp, &src, &interSrc, &interDst, &dst, &cropSize](cudaStream_t s) {
                reformatOp(s, src, interSrc);       // NCHW → NHWC
                op(s, interSrc, interDst, cropSize); // interleaved crop
                reformatOp(s, interDst, dst);       // NHWC → NCHW
            });
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor src = isPlanar
                             ? nvcv::Tensor({{srcShape.x, ch, srcShape.y, srcShape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{srcShape.x, srcShape.y, srcShape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{dstShape.x, ch, dstShape.y, dstShape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{dstShape.x, dstShape.y, dstShape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::CheckerboardValues<BT>());

        benchutils::warmup_and_exec(state, BENCH_CENTERCROP_WARMUP_ITERATIONS,
            [&op, &src, &dst, &cropSize](cudaStream_t s) { op(s, src, dst, cropSize); });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        throw std::invalid_argument("ImageBatchVarShape not implemented for this operator");
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(centercrop, NVBENCH_TYPE_AXES(BENCH_CENTERCROP_TYPES))
BENCH_CENTERCROP_AXES;

File: bench/python/ops/bench_centercrop.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA CenterCrop operator benchmark - Python equivalent of BenchCenterCrop.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_num_channels,
    get_dtype_size,
    create_tensor,
    create_stream_cache,
    run_benchmark,
)


def centercrop(state):
    """CenterCrop operator benchmark matching C++ BenchCenterCrop.cpp"""

    N, H, W = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    dtype = get_dtype(dtype_str)
    num_channels = get_num_channels(dtype_str)
    crop_type = state.get_string("cropType")
    input_kind = get_input_kind(state.get_string("inputKind"))
    try:
        layout = state.get_string("layout")
    except (KeyError, RuntimeError):
        layout = "NHWC"

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip(
            "CenterCrop benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts"
        )
        return None

    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"

    # CenterCrop is tensor-only; the native and fake planar paths are too.
    if input_kind == "VarShape":  # ImageBatchVarShape mode
        state.skip("ImageBatchVarShape not implemented for this benchmark")
        return None

    device_id = state.get_device()

    if crop_type == "SAME":
        crop_h, crop_w = H, W
    elif crop_type == "QUARTER":
        crop_h, crop_w = H // 2, W // 2
    else:
        state.skip(f"Invalid cropType: {crop_type}")
        return None

    dtype_size = get_dtype_size(dtype_str)
    full_bytes = N * H * W * dtype_size
    crop_bytes = N * crop_h * crop_w * dtype_size
    if is_fake_planar:
        # reformat(NCHW->NHWC) + crop + reformat(NHWC->NCHW)
        state.add_global_memory_reads(full_bytes + 2 * crop_bytes)
        state.add_global_memory_writes(full_bytes + 2 * crop_bytes)
    else:
        state.add_global_memory_reads(crop_bytes)
        state.add_global_memory_writes(crop_bytes)

    get_stream = create_stream_cache()

    if is_fake_planar:
        # Tensor-only "fake planar": planar->interleaved->crop->interleaved->planar,
        # all timed, as the comparison baseline for the native planar (NCHW) path.
        src = create_tensor(
            (N, num_channels, H, W),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode="checkerboard",
        )
        inter_src = create_tensor(
            (N, H, W, num_channels), dtype, device_id, layout="NHWC", fill_mode=0
        )
        inter_dst = create_tensor(
            (N, crop_h, crop_w, num_channels),
            dtype,
            device_id,
            layout="NHWC",
            fill_mode=0,
        )
        dst = create_tensor(
            (N, num_channels, crop_h, crop_w),
            dtype,
            device_id,
            layout="NCHW",
            fill_mode=0,
        )

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            cvcuda.center_crop_into(
                inter_dst, inter_src, (crop_w, crop_h), stream=stream
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    src_shape = (N, num_channels, H, W) if is_planar else (N, H, W, num_channels)
    dst_shape = (
        (N, num_channels, crop_h, crop_w)
        if is_planar
        else (N, crop_h, crop_w, num_channels)
    )
    src = create_tensor(
        src_shape, dtype, device_id, layout=layout, fill_mode="checkerboard"
    )
    dst = create_tensor(dst_shape, dtype, device_id, layout=layout, fill_mode=0)

    def run(launch):
        cvcuda.center_crop_into(dst, src, (crop_w, crop_h), stream=get_stream(launch))

    return run


if __name__ == "__main__":
    run_benchmark("centercrop", centercrop)
